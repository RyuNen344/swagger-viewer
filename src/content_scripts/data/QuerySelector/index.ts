import { getDocument } from "src/content_scripts/data/QuerySelector/Document"
import { FixMeAny } from "src/universal/types/utils"

/**
 * document.querySelectorAll() をモック可能にするためのラッパー
 */
export const querySelectorAll = (
  selector: string
): ReadonlyArray<HTMLElement> => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return Array.prototype.slice.call(getDocument().querySelectorAll(selector))
}
/**
 * 1要素をgetする
 *
 * @throws 2要素以上見つかった
 */
export const querySelector = (selector: string): HTMLElement | undefined => {
  const founds = querySelectorAll(selector)
  if (founds.length === 0) {
    return undefined
  }
  if (founds.length === 1) {
    return founds[0]
  }
  throw new Error(`Logic Failure: "${selector}" found multiple elements`)
}

/**
 * 必ず1要素をgetする
 *
 * @throws 見つからない | 2要素以上見つかった
 */
export const querySelectorStrict = <T extends HTMLElement>(
  selector: string
): T => {
  const founds = querySelectorAll(selector)
  if (founds.length === 1) {
    // なぜか型が合わないため無視
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return founds[0] as FixMeAny
  }
  throw new Error(`Logic Failure: "${selector}" is always 1 element exists`)
}
