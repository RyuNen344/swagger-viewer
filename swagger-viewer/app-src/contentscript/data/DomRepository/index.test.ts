import * as DomRepository from "."
import { createMockDocumentBy } from "../../__test__/Helper"
import * as Document from "../QuerySelector/Document"
import { html as swagger20JsonHtml } from "./fixtures/GitHubPageHtml_swagger_2_0_json"
import { html as swagger20YamlHtml_InPrivateBrowsing } from "./fixtures/GitHubPageHtml_swagger_2_0_yaml_InPrivateBrowsing"
import { html as swagger30YamlHtml } from "./fixtures/GitHubPageHtml_swagger_3_0_yaml"

describe("GitHubPageHtml swagger 2.0 json tests", () => {
  let sut: typeof DomRepository
  // let mockDocument: jest.MockInstance<typeof Document>
  let mockDocument: typeof Document

  beforeAll(() => {
    // ## Arrange ##
    jest.mock("../QuerySelector/Document")

    // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
    mockDocument = require("../QuerySelector/Document")
    // mock化しているため、mockReturnValueは必ず存在する
    // 型の書き方が不明なため、any castしている
    ;(mockDocument as any).getDocument.mockReturnValue(
      createMockDocumentBy(
        swagger20JsonHtml,
        "https://github.com/arx-8/swagger-viewer/blob/master/swagger-viewer/app-src/shared/__test__/fixtures/swagger_2_0.json",
      ),
    )

    // eslint-disable-next-line global-require
    sut = require(".")
  })

  describe("whole tests", () => {
    it("isAcceptableLocation", () => {
      // ## Assert ##
      expect(sut.isAcceptableLocation()).toEqual(true)
    })

    it("isConverted", () => {
      // ## Assert ##
      expect(sut.isConverted()).toEqual(false)
    })

    it("getElmOfSrcCode", () => {
      // ## Act ##
      const result = sut.getElmOfSrcCode()
      // ## Assert ##
      expect(result.textContent).not.toEqual(null)
      expect(result.children.length).toEqual(1)
    })

    it("extractSrc", () => {
      // ## Assert ##
      expect(sut.extractSrc()).toMatchSnapshot()
    })
  })
})

/**
 * プライベートブラウザ（シークレットモード） + organizationアカウント？では少し違ったDOMが表示されるため、別途テスト
 */
describe("GitHubPageHtml swagger 2.0 yaml tests in Private browsing", () => {
  let sut: typeof DomRepository
  // let mockDocument: jest.MockInstance<typeof Document>
  let mockDocument: typeof Document

  beforeAll(() => {
    // ## Arrange ##
    jest.mock("../QuerySelector/Document")

    // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
    mockDocument = require("../QuerySelector/Document")
    // mock化しているため、mockReturnValueは必ず存在する
    // 型の書き方が不明なため、any castしている
    ;(mockDocument as any).getDocument.mockReturnValue(
      createMockDocumentBy(
        swagger20YamlHtml_InPrivateBrowsing,
        "https://github.com/arx-8/swagger-viewer/blob/master/swagger-viewer/app-src/shared/__test__/fixtures/swagger_2_0.yaml",
      ),
    )

    // eslint-disable-next-line global-require
    sut = require(".")
  })

  describe("whole tests", () => {
    it("isAcceptableLocation", () => {
      // ## Assert ##
      expect(sut.isAcceptableLocation()).toEqual(true)
    })

    it("isConverted", () => {
      // ## Assert ##
      expect(sut.isConverted()).toEqual(false)
    })

    it("getElmOfSrcCode", () => {
      // ## Act ##
      const result = sut.getElmOfSrcCode()
      // ## Assert ##
      expect(result.textContent).not.toEqual(null)
      expect(result.children.length).toEqual(1)
    })

    it("extractSrc", () => {
      // ## Assert ##
      expect(sut.extractSrc()).toMatchSnapshot()
    })
  })
})

describe("GitHubPageHtml swagger 3.0 yaml tests", () => {
  let sut: typeof DomRepository
  // let mockDocument: jest.MockInstance<typeof Document>
  let mockDocument: typeof Document

  beforeAll(() => {
    // ## Arrange ##
    jest.mock("../QuerySelector/Document")

    // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
    mockDocument = require("../QuerySelector/Document")
    // mock化しているため、mockReturnValueは必ず存在する
    // 型の書き方が不明なため、any castしている
    ;(mockDocument as any).getDocument.mockReturnValue(
      createMockDocumentBy(
        swagger30YamlHtml,
        "https://github.com/arx-8/swagger-viewer/blob/master/swagger-viewer/app-src/shared/__test__/fixtures/swagger_3_0.yaml",
      ),
    )

    // eslint-disable-next-line global-require
    sut = require(".")
  })

  describe("whole tests", () => {
    it("isAcceptableLocation", () => {
      // ## Assert ##
      expect(sut.isAcceptableLocation()).toEqual(true)
    })

    it("isConverted", () => {
      // ## Assert ##
      expect(sut.isConverted()).toEqual(false)
    })

    it("getElmOfSrcCode", () => {
      // ## Act ##
      const result = sut.getElmOfSrcCode()
      // ## Assert ##
      expect(result.textContent).not.toEqual(null)
      expect(result.children.length).toEqual(1)
    })

    it("extractSrc", () => {
      // ## Assert ##
      expect(sut.extractSrc()).toMatchSnapshot()
    })
  })
})
