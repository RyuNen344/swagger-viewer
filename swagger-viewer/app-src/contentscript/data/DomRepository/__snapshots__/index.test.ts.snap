// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubPageHtml swagger 2.0 json tests whole tests extractSrc 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"version\\": \\"1.0.0\\",
    \\"title\\": \\"Swagger Petstore\\",
    \\"description\\": \\"A sample API that uses a petstore as an example to demonstrate features in the swagger-2.0 specification\\",
    \\"termsOfService\\": \\"http://swagger.io/terms/\\",
    \\"contact\\": {
      \\"name\\": \\"Swagger API Team\\"
    },
    \\"license\\": {
      \\"name\\": \\"MIT\\"
    }
  },
  \\"host\\": \\"petstore.swagger.io\\",
  \\"basePath\\": \\"/api\\",
  \\"schemes\\": [\\"http\\"],
  \\"consumes\\": [\\"application/json\\"],
  \\"produces\\": [\\"application/json\\"],
  \\"paths\\": {
    \\"/pets\\": {
      \\"get\\": {
        \\"description\\": \\"Returns all pets from the system that the user has access to\\",
        \\"operationId\\": \\"findPets\\",
        \\"produces\\": [
          \\"application/json\\",
          \\"application/xml\\",
          \\"text/xml\\",
          \\"text/html\\"
        ],
        \\"parameters\\": [
          {
            \\"name\\": \\"tags\\",
            \\"in\\": \\"query\\",
            \\"description\\": \\"tags to filter by\\",
            \\"required\\": false,
            \\"type\\": \\"array\\",
            \\"items\\": {
              \\"type\\": \\"string\\"
            },
            \\"collectionFormat\\": \\"csv\\"
          },
          {
            \\"name\\": \\"limit\\",
            \\"in\\": \\"query\\",
            \\"description\\": \\"maximum number of results to return\\",
            \\"required\\": false,
            \\"type\\": \\"integer\\",
            \\"format\\": \\"int32\\"
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"pet response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"$ref\\": \\"#/definitions/Pet\\"
              }
            }
          },
          \\"default\\": {
            \\"description\\": \\"unexpected error\\",
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorModel\\"
            }
          }
        }
      },
      \\"post\\": {
        \\"description\\": \\"Creates a new pet in the store.  Duplicates are allowed\\",
        \\"operationId\\": \\"addPet\\",
        \\"produces\\": [\\"application/json\\"],
        \\"parameters\\": [
          {
            \\"name\\": \\"pet\\",
            \\"in\\": \\"body\\",
            \\"description\\": \\"Pet to add to the store\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/NewPet\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"pet response\\",
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/Pet\\"
            }
          },
          \\"default\\": {
            \\"description\\": \\"unexpected error\\",
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorModel\\"
            }
          }
        }
      }
    },
    \\"/pets/{id}\\": {
      \\"get\\": {
        \\"description\\": \\"Returns a user based on a single ID, if the user does not have access to the pet\\",
        \\"operationId\\": \\"findPetById\\",
        \\"produces\\": [
          \\"application/json\\",
          \\"application/xml\\",
          \\"text/xml\\",
          \\"text/html\\"
        ],
        \\"parameters\\": [
          {
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"description\\": \\"ID of pet to fetch\\",
            \\"required\\": true,
            \\"type\\": \\"integer\\",
            \\"format\\": \\"int64\\"
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"pet response\\",
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/Pet\\"
            }
          },
          \\"default\\": {
            \\"description\\": \\"unexpected error\\",
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorModel\\"
            }
          }
        }
      },
      \\"delete\\": {
        \\"description\\": \\"deletes a single pet based on the ID supplied\\",
        \\"operationId\\": \\"deletePet\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"description\\": \\"ID of pet to delete\\",
            \\"required\\": true,
            \\"type\\": \\"integer\\",
            \\"format\\": \\"int64\\"
          }
        ],
        \\"responses\\": {
          \\"204\\": {
            \\"description\\": \\"pet deleted\\"
          },
          \\"default\\": {
            \\"description\\": \\"unexpected error\\",
            \\"schema\\": {
              \\"$ref\\": \\"#/definitions/ErrorModel\\"
            }
          }
        }
      }
    }
  },
  \\"definitions\\": {
    \\"Pet\\": {
      \\"type\\": \\"object\\",
      \\"allOf\\": [
        {
          \\"$ref\\": \\"#/definitions/NewPet\\"
        },
        {
          \\"required\\": [\\"id\\"],
          \\"properties\\": {
            \\"id\\": {
              \\"type\\": \\"integer\\",
              \\"format\\": \\"int64\\"
            }
          }
        }
      ]
    },
    \\"NewPet\\": {
      \\"type\\": \\"object\\",
      \\"required\\": [\\"name\\"],
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        },
        \\"tag\\": {
          \\"type\\": \\"string\\"
        }
      }
    },
    \\"ErrorModel\\": {
      \\"type\\": \\"object\\",
      \\"required\\": [\\"code\\", \\"message\\"],
      \\"properties\\": {
        \\"code\\": {
          \\"type\\": \\"integer\\",
          \\"format\\": \\"int32\\"
        },
        \\"message\\": {
          \\"type\\": \\"string\\"
        }
      }
    }
  }
}"
`;

exports[`GitHubPageHtml swagger 2.0 yaml tests in Private browsing whole tests extractSrc 1`] = `
"swagger: \\"2.0\\"
info:
  title: Simple API overview
  version: v2
paths:
  /:
    get:
      operationId: listVersionsv2
      summary: List API versions
      produces:
        - application/json
      responses:
        \\"200\\":
          description: |-
            200 300 response
          examples:
            application/json: |-
              {
                  \\"versions\\": [
                      {
                          \\"status\\": \\"CURRENT\\",
                          \\"updated\\": \\"2011-01-21T11:33:21Z\\",
                          \\"id\\": \\"v2.0\\",
                          \\"links\\": [
                              {
                                  \\"href\\": \\"http://127.0.0.1:8774/v2/\\",
                                  \\"rel\\": \\"self\\"
                              }
                          ]
                      },
                      {
                          \\"status\\": \\"EXPERIMENTAL\\",
                          \\"updated\\": \\"2013-07-23T11:33:21Z\\",
                          \\"id\\": \\"v3.0\\",
                          \\"links\\": [
                              {
                                  \\"href\\": \\"http://127.0.0.1:8774/v3/\\",
                                  \\"rel\\": \\"self\\"
                              }
                          ]
                      }
                  ]
              }
        \\"300\\":
          description: |-
            200 300 response
          examples:
            application/json: |-
              {
                  \\"versions\\": [
                      {
                          \\"status\\": \\"CURRENT\\",
                          \\"updated\\": \\"2011-01-21T11:33:21Z\\",
                          \\"id\\": \\"v2.0\\",
                          \\"links\\": [
                              {
                                  \\"href\\": \\"http://127.0.0.1:8774/v2/\\",
                                  \\"rel\\": \\"self\\"
                              }
                          ]
                      },
                      {
                          \\"status\\": \\"EXPERIMENTAL\\",
                          \\"updated\\": \\"2013-07-23T11:33:21Z\\",
                          \\"id\\": \\"v3.0\\",
                          \\"links\\": [
                              {
                                  \\"href\\": \\"http://127.0.0.1:8774/v3/\\",
                                  \\"rel\\": \\"self\\"
                              }
                          ]
                      }
                  ]
              }
  /v2:
    get:
      operationId: getVersionDetailsv2
      summary: Show API version details
      produces:
        - application/json
      responses:
        \\"200\\":
          description: |-
            200 203 response
          examples:
            application/json: |-
              {
                  \\"version\\": {
                      \\"status\\": \\"CURRENT\\",
                      \\"updated\\": \\"2011-01-21T11:33:21Z\\",
                      \\"media-types\\": [
                          {
                              \\"base\\": \\"application/xml\\",
                              \\"type\\": \\"application/vnd.openstack.compute+xml;version=2\\"
                          },
                          {
                              \\"base\\": \\"application/json\\",
                              \\"type\\": \\"application/vnd.openstack.compute+json;version=2\\"
                          }
                      ],
                      \\"id\\": \\"v2.0\\",
                      \\"links\\": [
                          {
                              \\"href\\": \\"http://127.0.0.1:8774/v2/\\",
                              \\"rel\\": \\"self\\"
                          },
                          {
                              \\"href\\": \\"http://docs.openstack.org/api/openstack-compute/2/os-compute-devguide-2.pdf\\",
                              \\"type\\": \\"application/pdf\\",
                              \\"rel\\": \\"describedby\\"
                          },
                          {
                              \\"href\\": \\"http://docs.openstack.org/api/openstack-compute/2/wadl/os-compute-2.wadl\\",
                              \\"type\\": \\"application/vnd.sun.wadl+xml\\",
                              \\"rel\\": \\"describedby\\"
                          },
                          {
                            \\"href\\": \\"http://docs.openstack.org/api/openstack-compute/2/wadl/os-compute-2.wadl\\",
                            \\"type\\": \\"application/vnd.sun.wadl+xml\\",
                            \\"rel\\": \\"describedby\\"
                          }
                      ]
                  }
              }
        \\"203\\":
          description: |-
            200 203 response
          examples:
            application/json: |-
              {
                  \\"version\\": {
                      \\"status\\": \\"CURRENT\\",
                      \\"updated\\": \\"2011-01-21T11:33:21Z\\",
                      \\"media-types\\": [
                          {
                              \\"base\\": \\"application/xml\\",
                              \\"type\\": \\"application/vnd.openstack.compute+xml;version=2\\"
                          },
                          {
                              \\"base\\": \\"application/json\\",
                              \\"type\\": \\"application/vnd.openstack.compute+json;version=2\\"
                          }
                      ],
                      \\"id\\": \\"v2.0\\",
                      \\"links\\": [
                          {
                              \\"href\\": \\"http://23.253.228.211:8774/v2/\\",
                              \\"rel\\": \\"self\\"
                          },
                          {
                              \\"href\\": \\"http://docs.openstack.org/api/openstack-compute/2/os-compute-devguide-2.pdf\\",
                              \\"type\\": \\"application/pdf\\",
                              \\"rel\\": \\"describedby\\"
                          },
                          {
                              \\"href\\": \\"http://docs.openstack.org/api/openstack-compute/2/wadl/os-compute-2.wadl\\",
                              \\"type\\": \\"application/vnd.sun.wadl+xml\\",
                              \\"rel\\": \\"describedby\\"
                          }
                      ]
                  }
              }
consumes:
  - application/json"
`;

exports[`GitHubPageHtml swagger 3.0 yaml tests whole tests extractSrc 1`] = `
"openapi: 3.0.0
info:
  title: Callback Example
  version: 1.0.0
paths:
  /streams:
    post:
      description: subscribes a client to receive out-of-band data
      parameters:
        - name: callbackUrl
          in: query
          required: true
          description: |
            the location where data will be sent.  Must be network accessible
            by the source server
          schema:
            type: string
            format: uri
            example: https://tonys-server.com
      responses:
        \\"201\\":
          description: subscription successfully created
          content:
            application/json:
              schema:
                description: subscription information
                required:
                  - subscriptionId
                properties:
                  subscriptionId:
                    description: this unique identifier allows management of the subscription
                    type: string
                    example: 2531329f-fb09-4ef7-887e-84e648214436
      callbacks:
        # the name \`onData\` is a convenience locator
        onData:
          # when data is sent, it will be sent to the \`callbackUrl\` provided
          # when making the subscription PLUS the suffix \`/data\`
          \\"{$request.query.callbackUrl}/data\\":
            post:
              requestBody:
                description: subscription payload
                content:
                  application/json:
                    schema:
                      properties:
                        timestamp:
                          type: string
                          format: date-time
                        userData:
                          type: string
              responses:
                \\"202\\":
                  description: |
                    Your server implementation should return this HTTP status code
                    if the data was received successfully
                \\"204\\":
                  description: |
                    Your server should return this HTTP status code if no longer interested
                    in further updates"
`;
